data_structures/treap.cpp

```diff
@@ -0,0 +1,228 @@
+     * @param x Erase from the subtree of node x (Usually x=root)
+     * @param k key
+     */
+    void _erase(int &x, int k) {
+        if (key[x] == k) {
+            if (cnt[x] > 1) {
+                cnt[x]--;
+            }  // If the node has more than one copy, the number of copies --
+            else {
+                if (childs[x][0] == 0 && childs[x][1] == 0) {
+                    x = 0;
+                    return;
+                }  // If there are no children, delete and return
+                // Otherwise, we need to rotate the sons and delete them
+                // recursively
+                int t = (priority[childs[x][0]] > priority[childs[x][1]]);
+                rotate(x, t);
+                _erase(x, k);
+            }
+        } else {  // Find the target value based on BST properties
+            _erase(childs[x][key[x] < k], k);
+        }
+        update(x);
+    }
+    /**
+     * @brief Find the KTH largest value (Also internal method)
+     * @param x Query the subtree of node x (Usually x=root)
+     * @param k The queried rank
+     * @return The element ranked x
+     */
+    int _getKth(int &x, int k) {
+        if (k <= size[childs[x][0]]) {
+            return _getKth(childs[x][0], k);
+        }
+        k -= size[childs[x][0]] + cnt[x];
+        if (k <= 0) {
+            return key[x];
+        }
+        return _getKth(childs[x][1], k);
+    }
+    /**
+     * @brief Query the rank of specified element (Also internal method)
+     * @param x Query the subtree of node x (Usually x=root)
+     * @param k The queried element
+     * @return The rank of element k
+     */
+    int _getRank(int x, int k) {
+        if (!x)
+            return 0;
+        if (k == key[x])
+            return size[childs[x][0]] + 1;
+        else if (k < key[x])
+            return _getRank(childs[x][0], k);
+        else
+            return size[childs[x][0]] + cnt[x] + _getRank(childs[x][1], k);
+    }
+    /**
+     * @brief Get the predecessor node of element k
+     * @param k The queried element
+     * @return The predecessor
+     */
+    int getPre(int k) {
+        int x = root, pre;
+        while (x) {
+            if (key[x] < k) {
+                pre = key[x], x = childs[x][1];
+            } else
+                x = childs[x][0];
+        }
+        return pre;
+    }
+    /**
+     * @brief Get the successor node of element k
+     * @param k The queried element
+     * @return The successor
+     */
+    int getNxt(int k) {
+        int x = root, nxt;
+        while (x) {
+            if (key[x] > k) {
+                nxt = key[x], x = childs[x][0];
+            } else
+                x = childs[x][1];
+        }
+        return nxt;
+    }
+    /**
+     * @brief External methods to set opertion object to be the whole tree
+     */
+    void insert(int k) { _insert(root, k); }
+    void erase(int k) { _erase(root, k); }
+    int getKth(int k) { return _getKth(root, k); }
```

Comment:
```suggestion
    int get_k_th(int k) { return _get_k_th(root, k); }
```
personal preference feel free to come up with a better and more descriptive function name